File: 2024/25/main.js
const { readFileSync } = require("fs");

const data = readFileSync("./input.txt", "utf-8").trim();

function processData() {
  const groups = data.split("\n\n").map(g => g.split("\n").map(r => r.split("")));
  const patterns = groups.map(g => {
    const counts = Array(g[0].length).fill(0);
    g.forEach(row => row.forEach((ch, i) => { if (ch === "#") counts[i]++; }));
    return { type: g[0][0] === "#" ? "lock" : "key", counts };
  });
  const locks = patterns.filter(p => p.type === "lock").map(p => p.counts);
  const keys = patterns.filter(p => p.type === "key").map(p => p.counts);
  let matchCount = 0;
  locks.forEach(lock => {
    keys.forEach(key => {
      if (lock.every((v, i) => v + key[i] <= 7)) matchCount++;
    });
  });
  return matchCount;
}

console.log(processData());

File: 2024/25/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func main() {
	data, _ := ioutil.ReadFile("./input.txt")
	lines := strings.TrimSpace(string(data))
	fmt.Println(processData(lines))
}

func processData(data string) int {
	groups := strings.Split(data, "\n\n")
	locks := [][]int{}
	keys := [][]int{}

	for _, group := range groups {
		rows := strings.Split(group, "\n")
		counts := make([]int, len(rows[0]))
		for _, row := range rows {
			for i, ch := range row {
				if ch == '#' {
					counts[i]++
				}
			}
		}
		if rows[0][0] == '#' {
			locks = append(locks, counts)
		} else {
			keys = append(keys, counts)
		}
	}

	matchCount := 0
	for _, lock := range locks {
		for _, key := range keys {
			if isValid(lock, key) {
				matchCount++
			}
		}
	}
	return matchCount
}

func isValid(lock, key []int) bool {
	for i := 0; i < len(lock); i++ {
		if lock[i]+key[i] > 7 {
			return false
		}
	}
	return true
}
